Fix incorrect check logic for state_forget_sig gate

jyoungyun opened the pull request:
State_forget_sig does not have Split operation as an input.
Instead it has Add operation and Add has a Split.

Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

Related issue : #21473

jyoungyun committed:
Fix incorrect check logic for state_forget_sig gate

State_forget_sig does not have Split operation as an input.
Instead it has Add operation and Add has a Split.

Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

jyoungyun commented:
I signed it!

drpngx commented:
@jyoungyun could you add a test perhaps? That would make it easier to make sure it's right.

tensorflowbutler commented:
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

jyoungyun committed:
Add lstm_test.py for testing LSTMCell operator

This commit checks if the LSTMCell is contained in model graph.

Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

jyoungyun commented:
@drpngx I added new patch for verifying my changes using python script. You can test this python script below commands.
```
$ python lstm_test.py --toco [toco location] --use_frozen_graph
```

When I tested this python script using original toco, it failed to find `LstmCell`:
```
tensorflow/contrib/lite/testing$ python lstm_test.py --toco [Original TOCO]/tensorflow/bazel-bin/tensorflow/contrib/lite/toco/toco --use_frozen_graph
========================
Not found 'LstmCell'
========================
```

But when I used this PR, the result was like below:
```
tensorflow/contrib/lite/testing$ python lstm_test.py --toco [applied this PR]/bazel-bin/tensorflow/contrib/lite/toco/toco --use_frozen_graph
========================
Found 'LstmCell'
========================
```

Also you can save the Graphviz dot or tflite file if you use `--save_dot` or `--save_tflite`.
Please take a look and leave comments.

jyoungyun commented:
Using original toco:
```
Number of all operator types: 7
	CONCATENATION                         :    2 	 (instrs: 0)
	FULLY_CONNECTED                       :    2 	 (instrs: 1,629,729,558,528)
	SPLIT                                 :    2 	 (instrs: ???)
	LOGISTIC                              :    5 	 (instrs: ???)
	TANH                                  :    4 	 (instrs: ???)
	MUL                                   :    5 	 (instrs: ???)
	ADD                                   :    2 	 (instrs: ???)
Number of all operators                       :   22 	 (total instrs: 1,629,729,558,528)
```

Using this PR:
```
Number of all operator types: 7
	CONCATENATION                         :    1 	 (instrs: 0)
	FULLY_CONNECTED                       :    1 	 (instrs: 814,864,779,264)
	SPLIT                                 :    1 	 (instrs: ???)
	LOGISTIC                              :    2 	 (instrs: ???)
	TANH                                  :    2 	 (instrs: ???)
	MUL                                   :    2 	 (instrs: ???)
	LSTM                                  :    1 	 (instrs: ???)
Number of all operators                       :   10 	 (total instrs: 814,864,779,264)
```

jyoungyun commented:
@drpngx Could you review this PR? Thanks in advance.

drpngx commented:
@aselle could you take another look?

tensorflowbutler commented:
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 44 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 59 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 74 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 89 days with no activity and the `awaiting review` label has been applied.

jyoungyun commented:
I rebased this pr.

tensorflowbutler commented:
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 15 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 30 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 45 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 60 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 75 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 89 days with no activity and the `awaiting review` label has been applied.

tensorflowbutler commented:
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 104 days with no activity and the `awaiting review` label has been applied.

jyoungyun commented:
@achowdhery @rthadur Do you have a plan to review this PR?

tensorflowbutler commented:
Nagging Reviewer @achowdhery: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

drpngx commented:
Could you please add another reviewer? I won't likely have time to do this
in the near future

On Wed, May 29, 2019, 9:28 AM Alfred Sorten Wolf <notifications@github.com>
wrote:

> Nagging Reviewer @achowdhery <https://github.com/achowdhery>: You have
> been added as a reviewer to this pull request. Please add your review or
> reassign. It has been 14 days with no activity and the awaiting review
> label has been applied.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/21474?email_source=notifications&email_token=AE75E3PWGADOZILHO6SQ5R3PX2AHLA5CNFSM4FOOPBCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWPJXDA#issuecomment-496933772>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE75E3JH4IJU6KSQRIHTK2TPX2AHLANCNFSM4FOOPBCA>
> .
>

ragmani commented on tensorflow/lite/toco/graph_transformations/identify_lstm.cc:
```diff
@@ -202,8 +202,14 @@ ::tensorflow::Status IdentifyLstmCell::Run(Model* model, std::size_t op_index,
+  Operator *state_forget_add;
+  if (!MatchOperatorInputs(*state_forget_sig, *model, OperatorType::kAdd,
+                           &state_forget_add)) {
+    return false;
```
Maybe `return ::tensorflow::Status::OK();`?
Anyway, this has helped me a lot.

ragmani commented on tensorflow/lite/toco/graph_transformations/identify_lstm.cc:
```diff
@@ -202,8 +202,14 @@ ::tensorflow::Status IdentifyLstmCell::Run(Model* model, std::size_t op_index,
-  if (!MatchOperatorInputs(*state_forget_sig, *model, OperatorType::kSplit,
-                           &tmp) ||
+  Operator *state_forget_add;
+  if (!MatchOperatorInputs(*state_forget_sig, *model, OperatorType::kAdd,
```
@jyoungyun It is correct for LSTM that "Spilt" comes before "forget sigmoid" without "Add".
BTW, my LSTM model using "static_rnn" also has the "Add". It's strange.

ragmani commented:
I have successfully converted to `LSTM` op without this patch by using `BasicRNNCell`.

miaout17 commented on tensorflow/contrib/lite/testing/lstm_test.py:
```diff
@@ -0,0 +1,274 @@
+import argparse
```
Instead of adding a whole new file, could you add the test case into the existing generated examples test?

https://github.com/tensorflow/tensorflow/blob/e0d9dfd54b0f2f4a2a6637268aa176d1701ee8e9/tensorflow/lite/testing/generate_examples_lib.py#L3320

The test is runnable by 
```
bazel test //tensorflow/lite/testing:zip_test_lstm
```

miaout17 commented on tensorflow/contrib/lite/testing/lstm_test.py:
```diff
@@ -0,0 +1,274 @@
+        print("========================")
+        if 'label=\"LstmCell\"' in model_binary.decode('utf-8'):
+          print("Found 'LstmCell'")
+        else:
```
It's not always safe to decode the model to UTF8. 
Doesn't the generated example test cover this?

miaout17 reviewed (changes requested):
I haven't noticed this pull request until now. 

@jyoungyun Sorry for the long review process. 
If you're still pursuing this pull request, could you rebase the change into tensorflow/lite/ (all the code was moved out of contrib/)

jyoungyun commented:
Sure thing. I will make a change soon.

On Fri, 28 Jun 2019 at 15:37 Yu-Cheng Ling <notifications@github.com> wrote:

> *@miaout17* requested changes on this pull request.
>
> I haven't noticed this pull request until now.
>
> @jyoungyun <https://github.com/jyoungyun> Sorry for the long review
> process.
> If you're still pursuing this pull request, could you rebase the change
> into tensorflow/lite/ (all the code was moved out of contrib/)
> ------------------------------
>
> In tensorflow/contrib/lite/testing/lstm_test.py
> <https://github.com/tensorflow/tensorflow/pull/21474#discussion_r298466793>
> :
>
> > @@ -0,0 +1,274 @@
> +import argparse
>
> Instead of adding a whole new file, could you add the test case into the
> existing generated examples test?
>
>
> https://github.com/tensorflow/tensorflow/blob/e0d9dfd54b0f2f4a2a6637268aa176d1701ee8e9/tensorflow/lite/testing/generate_examples_lib.py#L3320
>
> The test is runnable by
>
> bazel test //tensorflow/lite/testing:zip_test_lstm
>
> ------------------------------
>
> In tensorflow/contrib/lite/testing/lstm_test.py
> <https://github.com/tensorflow/tensorflow/pull/21474#discussion_r298467367>
> :
>
> > +        input_tensors = [(input_tensor.name.split(":")[0],
> +                          input_tensor.get_shape(), input_tensor.dtype)
> +                         for input_tensor in inputs]
> +        output_tensors = [normalize_output_name(out.name) for out in outputs]
> +
> +        graph_def = freeze_graph(
> +            sess,
> +            tf.global_variables() + inputs + outputs) if FLAGS.use_frozen_graph else sess.graph_def
> +        model_binary = toco_convert(
> +            graph_def.SerializeToString(), input_tensors, output_tensors)
> +
> +        # Check if `LstmCell` is contained
> +        print("========================")
> +        if 'label=\"LstmCell\"' in model_binary.decode('utf-8'):
> +          print("Found 'LstmCell'")
> +        else:
>
> It's not always safe to decode the model to UTF8.
> Doesn't the generated example test cover this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/21474?email_source=notifications&email_token=ABXDSS4LSMWKYOM5MZAJAULP4WWRFA5CNFSM4FOOPBCKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB46CWIY#pullrequestreview-255601443>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABXDSSZMNXVBXWKCRFGSFD3P4WWRFANCNFSM4FOOPBCA>
> .
>

mellanox-github commented:
Can one of the admins verify this patch?

drpngx commented:
Hey guys, it's best if someone else reviews this. It seems that it's fine
in principle. @yifeif do you want to take a crack at it?

On Tue, Aug 20, 2019 at 6:30 AM Mellanox Build Bot <notifications@github.com>
wrote:

> Can one of the admins verify this patch?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/21474?email_source=notifications&email_token=AE75E3MTTGJ3BX56DFZPSC3QFPWXNA5CNFSM4FOOPBCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4WIELQ#issuecomment-523010606>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE75E3MR2RJGSFTHJQGFVCTQFPWXNANCNFSM4FOOPBCA>
> .
>

miaout17 commented:
@drpngx I think we're waiting for the author to address the comments.

rthadur commented:
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
